packages/tldraw/src/lib/ui/hooks/useFlatten.ts

```diff
@@ -0,0 +1,191 @@
+		const expandedBounds = ids.map((id) => {
+			return {
+				id,
+				bounds: editor.getShapeMaskedPageBounds(id)!.clone().expandBy(multipleImagePadding),
+			}
+		})
+
+		for (let i = 0; i < expandedBounds.length; i++) {
+			const item = expandedBounds[i]
+			if (i === 0) {
+				groups[0] = {
+					ids: [item.id],
+					bounds: item.bounds,
+				}
+				continue
+			}
+
+			let didLand = false
+
+			for (const group of groups) {
+				if (group.bounds.includes(item.bounds)) {
+					group.ids.push(item.id)
+					group.bounds.expand(item.bounds)
+					didLand = true
+					break
+				}
+			}
+
+			if (!didLand) {
+				groups.push({
+					ids: [item.id],
+					bounds: item.bounds,
+				})
+			}
+		}
+	} else {
+		const bounds = Box.Common(ids.map((id) => editor.getShapeMaskedPageBounds(id)!))
+		groups.push({
+			ids,
+			bounds,
+		})
+	}
+
+	const padding = editor.options.flattenImageBoundsPadding
+
+	for (const group of groups) {
+		if (multipleImagePadding !== undefined) {
+			// shrink the bounds again, removing the expanded area
+			group.bounds.expandBy(-multipleImagePadding)
+		}
+
+		// get an image for the shapes
+		const svgResult = await editor.getSvgString(group.ids, {
+			padding,
+		})
+		if (!svgResult?.svg) continue
+
+		// get an image asset for the image
+		const blob = new Blob([svgResult.svg], { type: 'image/svg+xml' })
+		const asset = (await editor.getAssetForExternalContent({
+			type: 'file',
+			file: new File([blob], 'asset.svg', { type: 'image/svg+xml' }),
+		})) as TLImageAsset
+		if (!asset) continue
+
+		// add it to the group
+		group.asset = asset
+	}
+
+	const shapeIds: TLShapeId[] = []
+
+	transact(() => {
+		for (const group of groups) {
+			const { asset, bounds, ids } = group
+			if (!asset) continue
+
+			const assetId = AssetRecordType.createId()
+
+			const commonAncestorId = editor.findCommonAncestor(ids) ?? editor.getCurrentPageId()
+			if (!commonAncestorId) continue
+
+			let index: IndexKey = 'a1' as IndexKey
+			for (const id of ids) {
+				const shape = editor.getShape(id)
+				if (!shape) continue
+				if (shape.parentId === commonAncestorId) {
+					if (shape.index > index) {
+						index = shape.index
+					}
+					break
+				}
+			}
+
+			let x: number
+			let y: number
+			let rotation: number
+
+			if (isShapeId(commonAncestorId)) {
+				const commonAncestor = editor.getShape(commonAncestorId)
+				if (!commonAncestor) continue
+				// put the point in the parent's space
+				const point = editor.getPointInShapeSpace(commonAncestor, {
+					x: bounds.x,
+					y: bounds.y,
+				})
+				// get the parent's rotation
+				rotation = editor.getShapePageTransform(commonAncestorId).rotation()
+				// rotate the point against the parent's rotation
+				point.sub(new Vec(padding, padding).rot(-rotation))
+				x = point.x
+				y = point.y
+			} else {
+				// if the common ancestor is the page, then just adjust for the padding
+				x = bounds.x - padding
+				y = bounds.y - padding
+				rotation = 0
+			}
+
```

Comment:
add a `mark` here to for undo/redo

Reply:
I've marked in the action, which feels like the right place for it

Reply:
mm, something's not right about it - I can undo but I can't redo unless there's an explicit `mark`. Unless I'm missing something about how new history stuff works 🤔